source/common/upstream/outlier_detection_impl.h

```diff
@@ -99,50 +101,107 @@ class SuccessRateAccumulator {
 
   void eject(MonotonicTime ejection_time);
   void uneject(MonotonicTime ejection_time);
-  void updateCurrentSuccessRateBucket();
-  SuccessRateAccumulator& successRateAccumulator() { return success_rate_accumulator_; }
-  void successRate(double new_success_rate) { success_rate_ = new_success_rate; }
+
   void resetConsecutive5xx() { consecutive_5xx_ = 0; }
   void resetConsecutiveGatewayFailure() { consecutive_gateway_failure_ = 0; }
+  void resetConsecutiveLocalOriginFailure() { consecutive_local_origin_failure_ = 0; }
   static Http::Code resultToHttpCode(Result result);
 
   // Upstream::Outlier::DetectorHostMonitor
   uint32_t numEjections() override { return num_ejections_; }
   void putHttpResponseCode(uint64_t response_code) override;
-  void putResult(Result result) override;
+  void putResult(Result result, absl::optional<uint64_t> code) override;
   void putResponseTime(std::chrono::milliseconds) override {}
   const absl::optional<MonotonicTime>& lastEjectionTime() override { return last_ejection_time_; }
   const absl::optional<MonotonicTime>& lastUnejectionTime() override {
     return last_unejection_time_;
   }
-  double successRate() const override { return success_rate_; }
+
+  // Template based and function based methods for managing SuccessRate detectors
+  template <envoy::data::cluster::v2alpha::OutlierEjectionType Type>
+  const std::unique_ptr<SuccessRateMonitor>& getSRMonitor() const {
+    return success_rate_monitors_.at(Type);
+  }
+  const std::unique_ptr<SuccessRateMonitor>&
+  getSRMonitor(envoy::data::cluster::v2alpha::OutlierEjectionType Type) const {
+    return success_rate_monitors_.at(Type);
+  }
+
+  double successRate(envoy::data::cluster::v2alpha::OutlierEjectionType type) const override {
+    return getSRMonitor(type)->getSuccessRate();
+  }
+  void updateCurrentSuccessRateBucket();
+  void successRate(envoy::data::cluster::v2alpha::OutlierEjectionType type,
+                   double new_success_rate) {
+    getSRMonitor(type)->setSuccessRate(new_success_rate);
+  }
+
+  // handlers for reporting local origin errors
+  void localOriginFailure();
+  void localOriginNoFailure();
 
 private:
   std::weak_ptr<DetectorImpl> detector_;
   std::weak_ptr<Host> host_;
-  std::atomic<uint32_t> consecutive_5xx_{0};
-  std::atomic<uint32_t> consecutive_gateway_failure_{0};
   absl::optional<MonotonicTime> last_ejection_time_;
   absl::optional<MonotonicTime> last_unejection_time_;
   uint32_t num_ejections_{};
-  SuccessRateAccumulator success_rate_accumulator_;
-  std::atomic<SuccessRateAccumulatorBucket*> success_rate_accumulator_bucket_;
-  double success_rate_;
+
+  // counters for externally generated failures
+  std::atomic<uint32_t> consecutive_5xx_{0};
+  std::atomic<uint32_t> consecutive_gateway_failure_{0};
+
+  // counters for local origin failures
+  std::atomic<uint32_t> consecutive_local_origin_failure_{0};
+
+  absl::flat_hash_map<envoy::data::cluster::v2alpha::OutlierEjectionType,
```

Comment:
Do we need the hash map and related complexity? It seems like you can just have 2 stack allocated SR monitors and then a function that returns one of them depending on the config?

Reply:
Yes. Probably it was overkill. Allocated 2 instances of SR monitors and added simple method to return one of them depending on the type: https://github.com/cpakulski/envoy/blob/issue/3643/source/common/upstream/outlier_detection_impl.h#L158-L160